src/executor.cc

```diff
@@ -111,6 +111,14 @@ Result Executor::Execute(Engine* engine,
       r = engine->DoClearDepth(cmd->AsClearDepth());
     } else if (cmd->IsClearStencil()) {
       r = engine->DoClearStencil(cmd->AsClearStencil());
+    } else if (cmd->IsCopy()) {
+      auto copy = cmd->AsCopy();
+      if (copy->GetBufferTo()->GetMemPtr() != nullptr) {
+        // FIXME: clean the buffer?
+        copy->GetBufferTo()->SetMemPtr(nullptr);
+      }
+      std::vector<Value> data = copy->GetBufferFrom()->GetData();
+      copy->GetBufferTo()->SetData(std::move(data));
```

Comment:
I think we want to put this method into buffer itself, then executor just does `Result r = buffer_from->CopyTo(buffer_to);`

Then in the buffer code you need to do something like:

````
// TODO(hevrard) Return error if buffer_to mem_ptr_ is backed by GPU memory
free(buffer_to->mem_ptr_);
buffer_to_->values_.clear();
bufer_to_values_ = values_;
if (mem_ptr_ != nullptr) {
  malloc(buffer_to->mem_ptr_, width * height * format byte size)
  std::memcpy(buffer_to->mem_ptr_, ....);
}
return {}
```

The GetMemPtr is a bit tricky as it returns either the mem_ptr_ or the values data buffer, and so we can't use it directly to do the copy.

This also means you'll need to be sure we're cleaning up mem_ptr_ correctly in this case. I think I assume'd it would get cleaned up through the GPU code?

Reply:
Yeah I spotted that GetMemPtr is a bit tricky. Thanks for the guidance here.

So a buffer stores things in either data_, values_, or mem_ptr_? Is this exclusive, i.e., only one of these three is allowed at a time?

Reply:
To be on the safe side, I'd just copy all of them. In general only one should be set, but if we don't need to make that assumption, I'd prefer not too.